specs/SetMemoryUsageLevel.md

```diff
@@ -0,0 +1,160 @@
+}
+
+void ViewComponent::OnBecomingActive()
+{
+  BOOL isSuspended = FALSE;
+  CHECK_FAILURE(m_webview->get_IsSuspended(&isSuspended));
+  if (isSuspended)
+  {
+     CHECK_FAILURE(m_webView->Resume());
+  }
+  else
+  {
+     COREWEBVIEW2_MEMORY_USAGE_TARGET_LEVEL memoryUsageTargetLevel = COREWEBVIEW2_MEMORY_USAGE_TARGET_LEVEL_LOW;
+     CHECK_FAILURE(m_webview->get_MemoryUsageTargetLevel(&memoryUsageTargetLevel));
+     if (memoryUsageTargetLevel == COREWEBVIEW2_MEMORY_USAGE_LEVEL_LOW)
+     {
+         CHECK_FAILURE(m_webView->put_MemoryUsageTargetLevel(COREWEBVIEW2_MEMORY_USAGE_TARGET_LEVEL_NORMAL));
+     }
+  }
+}
+```
+
+# Remarks
+See [API Details](#api-details) section below for more details.
+
+# API Notes
+See [API Details](#api-details) section below for API reference.
+
+# API Details
+
+## Win32 C++
+```IDL
+/// Specifies memory usage target level of WebView.
+[v1_enum]
+typedef enum COREWEBVIEW2_MEMORY_USAGE_TARGET_LEVEL
+{
+    /// Specifies normal memory usage target level.
+    COREWEBVIEW2_MEMORY_USAGE_TARGET_LEVEL_NORMAL,
+
+    /// Specifies low memory usage target level.
+    /// Used for inactivate WebView for reduced memory consumption.
+    COREWEBVIEW2_MEMORY_USAGE_TARGET_LEVEL_LOW,
+
+} COREWEBVIEW2_MEMORY_USAGE_LEVEL;
+
+interface ICoreWebView2_6 : ICoreWebView2
+{
+
+  /// `MemoryUsageTargetLevel` indicates desired memory comsumption level of WebView.
+  HRESULT get_MemoryUsageTargetLevel([in] COREWEBVIEW2_MEMORY_USAGE_TARGET_LEVEL* level);
+  
+  /// An app may set `MemoryUsageTargetLevel` to indicate desired memory
+  /// comsumption level of WebView. Scripts will not be impacted and continue to run.
+  /// This is useful for inactive apps that still want to run scripts and/or keep
+  /// network connections alive and therefore could not call `TrySuspend` and `Resume`
+  /// to reduce memory consumption.
+  /// These apps can set memory usage target level to `COREWEBVIEW2_MEMORY_USAGE_TARGET_LEVEL_LOW` when
+  /// the app becomes inactive, and set back to `COREWEBVIEW2_MEMORY_USAGE_TARGET_LEVEL_NORMAL` when
+  /// the app becomes active.
+  /// It is not neccesary to set CoreWebView2Controller's IsVisible property to false when calling the API.
+  /// It is a best effort operation to change memory usage level, and the API will return before the operation completes.
+  /// Setting the level to `COREWEBVIEW2_MEMORY_USAGE_TARGET_LEVEL_LOW` could potentially cause
+  /// memory for some WebView browser processes to be swapped out to disk when needed. Therefore,
+  /// it is important for the app to set the level back to `COREWEBVIEW2_MEMORY_USAGE_TARGET_LEVEL_NORMAL`
+  /// when the app becomes active again to have a smooth user experience.
+  /// Setting memory usage level back to normal will not happen automatically.
+  /// An app should choose to use either the combination of `TrySuspend` and `Resume` or the combination
+  /// of setting MemoryUsageTargetLevel to low and normal. It is not advisable to mix them.
```

Comment:
If it's not advisable to mix, should we block it? The details in the next few lines about how they interact is getting complex. It's confusing that Suspend() sets the memory level, since I assumed Suspend() would suspend and so there's no issue of memory level (memory level is supposed to be about allowing scripts still run).

Rather than a property, could this be a parameter on Suspend()? Like a mode where you're not fully suspended but in a low-memory constraint.

Reply:
We were actually struggling between a new API or overloading TrySuspend.

"Suspend" suggests reducing CPU consumption and the new thing is related to memory only. So, it feels strange for us to add a Suspend level that doesn't impact CPU consumption. That's the main reason we didn't make it a parameter of TrySuspend. Both are for similar scenario though.
Other considerations:

-  TrySuspend currently requires WebView2Controller to be invisible while reduce memory usage doesn't have this requirement.

- For WebView2 C++ API, there is no overload support so we would need a new method name like TrySuspendToLevel(level) and another property to know the suspension level. For .NET and WinRT, we would want to overload TrySuspend and then we got different API names different languages.

If we all feel that a parameter to TrySuspend is a better design, we could redesign it.

Reply:
For whether we should block the mix usage of the 2 sets of the APIs. There could be rare but valid app scenario where the app first set memory usage level to low and still run some script, and then it doesn't need run script anymore and call Suspend. If could require the app to go back to normal memory usage and then suspend. That would cause more churns though. If we all feel that it is easier for developer to understand, we could block setting memory level when suspended and/or blocking suspend/resume when memory level is not normal.
You are right, when suspended there is no reason to consume a lot of memory. WebView2 Suspend is not like UWP suspend and currently does not suspend the processes at OS level though. So, internally it will set memory level to low for suspend. The question is how we want developers to perceive this new property. That would decide what level to report when suspended. We feel that reporting as low when suspended feels more natural.

Reply:
If no strong objection, I would like to:

- Keep this as a separate property.
- Block setting it if already suspended.

Reply:
Updated to block setting MemoryUsageTargetLevel while suspended.

Reply:
For TrySuspend I was thinking maybe there could be two modes, where you could choose to suspend memory or suspend CPU. But seeing that written, I think that does sound odd. I'm not sure what you mean about C++ API not supporting overloads though? WinRT has a policy against type-based overloads, but arity-overloads (different number of parameters) is OK.

When you say block the property do you mean throw an exception? Properties that throw exceptions are frustrating. Preferable would be to ignore the property, unless/until a Resume() happens.

Reply:
Will update to "Trying to set `MemoryUsageTargetLevel` while suspended will be ignored".